### PR TITLE
REF: Remove DatetimelikeArrayMixin._shallow_copy

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken off of #23426.

### CODE REVIEW

**Comment by gfyoung** [comment]
A couple of points:

* Any performance impact with this refactor?
* For the less initiated, why don't we want to use `_shallow_copy` ?

**Comment by jbrockmendel** [comment]
> Any performance impact with this refactor?

If anything, it removes a level of function-call overhead.  Should not make a measurable impact.

> For the less initiated, why don't we want to use `_shallow_copy` ?

_shallow_copy was implemented in these EA mixin classes to make them play nicely when inherited by their respective Index subclasses.  We are moving away from having these get mixed in, and do not need _shallow_copy in the standalone EA subclasses.

**Comment by jreback** [comment]
thanks

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -586,7 +586,7 @@ def tz_convert(self, tz):
                             'tz_localize to localize')
 
         # No conversion since timestamps are all UTC to begin with
-        return self._shallow_copy(tz=tz)
+        return self._simple_new(self.asi8, tz=tz, freq=self.freq)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: `_simple_new(int64values, tz=tz)` will get you the same thing as `__new__(int64values, tz=tz)`, but the same is _not_ true with M8 values.  I find that really helpful.

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimes.py]: You are referring to the timezone difference you mentioned earlier?

I personally find that very confusing that `DatetimeArray(self._data, tz=..)` and `DatetimeArray(self._data.view(int), tz=..)` do not give the same. To the extent possible, I would not repeat this pattern for our Arrays? 
(of course, in the DatetimeIndex constructor we cannot just change it, unless we want to deprecate that first).

**Pull request merged**